Linux Page Cache and IDR

gcmoreira opened the pull request:
This PR includes abstractions for RadixTree, XArray, Page Cache and IDR to ensure support with both older and the newest kernel versions. The RadixTree and XArray enable the implementation of other subsystems like the Page Cache and IDR.

This is the final iteration of other PRs which are part of this feature. 
- https://github.com/volatilityfoundation/volatility3/pull/1222
- https://github.com/volatilityfoundation/volatility3/pull/1230


Besides the Page Cache plugins, we added two more plugins to this PR to be able to test the whole code with different kernel versions.

# Inode Page Cache
## linux.pagecache.Files plugin 
This plugin can list all inodes, filter by type and find filenames. Basically, it shows superblock, dentry and inode info.
It also supports the Timeliner plugin.

For instance, to list all inodes:
```shell
$ ./vol.py -f ./ubuntu-4.9.0-19-32bit.core linux.pagecache.Files 
Volatility 3 Framework 2.7.1
Progress:  100.00               Stacking attempts finished                  
SuperblockAddr  MountPoint      Device  InodeNum        InodeAddr       FileType        InodePages      CachedPages     FileMode        AccessTime      ModificationTime        ChangeTime      FilePath

0xf7094800      /       0:1     1       0xf4801e60      DIR     0       0       drwxr-xr-x      2024-07-29 13:44:58.916000 UTC  2024-07-29 13:44:58.916000 UTC  2024-07-29 13:44:58.916000 UTC  /
0xf7094800      /       0:1     7626    0xf4b2a5c0      DIR     0       0       drwx------      2022-06-30 18:26:10.000000 UTC  2022-06-30 18:26:10.000000 UTC  2024-07-29 13:44:58.228996 UTC  /root
0xf7154800      /sys    0:16    1       0xf4b492e0      DIR     0       0       dr-xr-xr-x      2024-07-29 13:44:58.500000 UTC  2024-07-29 13:44:58.500000 UTC  2024-07-29 13:44:58.500000 UTC  /sys
...
```
Filtering by file type. For instance, list the top Regular files with more pages in cache:
```shell
$ ./vol.py -f ./ubuntu-4.9.0-19-32bit.core linux.pagecache.Files --type REG | sort -k8 -n -r | head -10
0xf6845800      /run    0:18    9964    0xf6fd30b0      REG     647     647     -rw-r-----      2024-07-29 13:45:01.348000 UTC  2024-07-29 13:45:02.044000 UTC  2024-07-29 13:45:02.044000 UTC  /run/log/journal/791e6a5cbadf47da9808b2fdd26a7bf0/system.journal
0xf6f7b800      /       254:3   5506492 0xf46d5780      REG     549     474     -rwxr-xr-x      2024-07-29 13:00:30.392000 UTC  2020-06-20 17:33:46.000000 UTC  2024-01-07 10:46:57.080673 UTC  /usr/bin/perl
0xf6f7b800      /       254:3   2359824 0xf4536a08      REG     517     440     -rw-r--r--      2024-07-29 13:00:29.524000 UTC  2022-06-29 12:41:17.000000 UTC  2024-01-07 10:56:29.475108 UTC  /lib/systemd/libsystemd-shared-232.so
...
```

Find a specific file name:
```shell
$ ./vol.py -f ./ubuntu-4.9.0-19-32bit.core linux.pagecache.Files  --find /etc/shadow
Volatility 3 Framework 2.7.1
Progress:  100.00               Stacking attempts finished                  
SuperblockAddr  MountPoint      Device  InodeNum        InodeAddr       FileType        InodePages      CachedPages     FileMode        AccessTime      ModificationTime        ChangeTime      FilePath

0xf6f7b800      /       254:3   7078755 0xf46c83f8      REG     1       1       -rw-r-----      2024-07-29 13:00:30.552000 UTC  2024-01-07 11:19:42.597348 UTC  2024-01-07 11:19:42.597348 UTC  /etc/shadow

```
## linux.pagecache.InodePages plugin 

This plugin requires either an inode address (--inode) or a filename (--find) to search for the inode, and optionally, you can use (--dump) to dump the inode's page cache contents which is the primary and greatest goal of this effort.

Example 1, via inode address:
```shell
$ ./vol.py -f ./ubuntu-4.9.0-19-32bit.core linux.pagecache.InodePages --inode 0xf6fd30b0 --dump system.journal
Volatility 3 Framework 2.7.1
PageVAddr       PagePAddr       MappingAddr     Index   DumpSafe        Flags
 
0xf5e9be90      0x7bf84000      0xf6fd31a8      0       True    active,dirty,lru,referenced,savepinned,swapbacked,uptodate
0xf5e9beb4      0x7bf85000      0xf6fd31a8      1       True    active,dirty,lru,referenced,savepinned,swapbacked,uptodate
0xf5e9bed8      0x7bf86000      0xf6fd31a8      2       True    active,dirty,lru,referenced,savepinned,swapbacked,uptodate
...
0xf5e8f944      0x7ba09000      0xf6fd31a8      646     True    dirty,lru,savepinned,swapbacked
```

Same inode but finding it via its filename:
```shell
$ ./vol.py -f ./ubuntu-4.9.0-19-32bit.core linux.pagecache.InodePages --find /run/log/journal/791e6a5cbadf47da9808b2fdd26a7bf0/system.journal --dump system.journal
Volatility 3 Framework 2.7.1
PageVAddr       PagePAddr       MappingAddr     Index   DumpSafe        Flags
 
0xf5e9be90      0x7bf84000      0xf6fd31a8      0       True    active,dirty,lru,referenced,savepinned,swapbacked,uptodate
0xf5e9beb4      0x7bf85000      0xf6fd31a8      1       True    active,dirty,lru,referenced,savepinned,swapbacked,uptodate
0xf5e9bed8      0x7bf86000      0xf6fd31a8      2       True    active,dirty,lru,referenced,savepinned,swapbacked,uptodate
...
0xf5e8f944      0x7ba09000      0xf6fd31a8      646     True    dirty,lru,savepinned,swapbacked
```
Checking the results:
```shell
$ file system.journal 
system.journal: Journal file, Mon Jul 29 13:44:59 2024, online, compressed lz4, header size 0xf0, entries 0x23f

$ journalctl --file=system.journal --list-boots
IDX BOOT ID                          FIRST ENTRY                  LAST ENTRY                  
  0 4f2a7426dab64f4dafa29343d13c1483 Mon 2024-07-29 23:44:59 AEST Mon 2024-07-29 23:45:01 AEST

$ journalctl --file=system.journal 
Jul 29 23:44:59 debian9.localdomain kernel: Linux version 4.9.0-19-686 (debian-kernel@lists.debian.org) (gcc version 6.3.0 20170516 (Debian 6.3.0-18+deb9u1) ) #1 SMP Debian 4.9.320-2 (2022-06-30)
Jul 29 23:44:59 debian9.localdomain kernel: x86/fpu: Supporting XSAVE feature 0x001: 'x87 floating point registers'
Jul 29 23:44:59 debian9.localdomain kernel: x86/fpu: Supporting XSAVE feature 0x002: 'SSE registers'
Jul 29 23:44:59 debian9.localdomain kernel: x86/fpu: Supporting XSAVE feature 0x004: 'AVX registers'
...
```
Example 2:
```shell
$ ./vol.py ./ubuntu-4.9.0-19-686.core linux.pagecache.Files --find /etc/passwd
Volatility 3 Framework 2.7.1        
SuperblockAddr  MountPoint      Device  InodeNum        InodeAddr       FileType        InodePages      CachedPages     FileMode        AccessTime      ModificationTime        ChangeTime      FilePath

0xf6f7b800      /       254:3   7078938 0xf45116c0      REG     1       1       -rw-r--r--      2024-07-29 13:00:30.368000 UTC  2024-01-07 11:07:28.924695 UTC  2024-01-07 11:07:28.924695 UTC  /etc/passwd
```
Dump the inode:
```shell
$ ./vol.py -f ./ubuntu-4.9.0-19-686.core linux.pagecache.InodePages --find /etc/passwd --dump passwd
Volatility 3 Framework 2.7.1
PageVAddr       PagePAddr       MappingAddr     Index   DumpSafe        Flags

0xf5e87b68      0x7b68a000      0xf45117b8      0       True    active,lru,mappedtodisk,referenced,uptodate
```

Check the file:
```shell
$ cat passwd
root:x:0:0:root:/root:/bin/bash
daemon:x:1:1:daemon:/usr/sbin:/usr/sbin/nologin
bin:x:2:2:bin:/bin:/usr/sbin/nologin
sys:x:3:3:sys:/dev:/usr/sbin/nologin
sync:x:4:65534:sync:/bin:/bin/sync
games:x:5:60:games:/usr/games:/usr/sbin/nologin
man:x:6:12:man:/var/cache/man:/usr/sbin/nologi
...
```

# IDR
## linux.ebpf.EBPF
A basic plugin to enumerate eBPF programs using the prog IDR. Tip for devs: Other contributions using the maps, links, etc. IDRs and even extending this plugin are welcome ;)

Test the plugin on a Cilium installation.
```shell
$ ./vol.py -f ./ram-5.8.0-53-ubuntu20.04-64bit.lime_cilium linux.ebpf.EBPF 
Volatility 3 Framework 2.8.0
Progress:  100.00               Stacking attempts finished                 
Address Name    Tag     Type

0xc9000002f000  -       3b185187f1855c4c        BPF_PROG_TYPE_XDP
0xc90000031000  BEGIN   f4b3d30cab782162        BPF_PROG_TYPE_KPROBE
0xc90000067000  sys_enter_execv e4476dee689a817e        BPF_PROG_TYPE_TRACEPOINT
0xc90000063000  -       3b185187f1855c4c        BPF_PROG_TYPE_XDP
0xc900005d5000  xdp_prog_simple 3b185187f1855c4c        BPF_PROG_TYPE_XDP
0xc90000141000  -       59904229c5a1f55d        BPF_PROG_TYPE_SCHED_CLS
0xc90000559000  -       6deef7357e7b4530        BPF_PROG_TYPE_CGROUP_SKB
0xc90000409000  -       6deef7357e7b4530        BPF_PROG_TYPE_CGROUP_SKB
0xc9000079d000  -       6deef7357e7b4530        BPF_PROG_TYPE_CGROUP_SKB
0xc9000079b000  -       6deef7357e7b4530        BPF_PROG_TYPE_CGROUP_SKB
...
```
## linux.pidhashtable.PIDHashTable
This is based on the vol2 plugin, removing ancient kernel support, curating code and enhancing comments, while using the new IDR abstraction included also in this effort.

```shell
$ ./vol.py -f ./ram-5.8.0-53-ubuntu20.04-64bit.lime_cilium linux.pidhashtable.PIDHashTable --decorate-comm 
Volatility 3 Framework 2.8.0
Progress:  100.00               Stacking attempts finished                 
OFFSET  PID     TID     PPID    COMM

0x8881396117c0  1       1       0       systemd
0x888139615f00  2       2       0       [kthreadd]
0x888139614740  3       3       2       [rcu_gp]
0x888139610000  4       4       2       [rcu_par_gp]
0x8881396317c0  6       6       2       [kworker/0:0H]
0x888139630000  9       9       2       [mm_percpu_wq]
0x888139632f80  10      10      2       [ksoftirqd/0]
0x8881396397c0  11      11      2       [rcu_sched]
...
```

gcmoreira committed:
Linux: Add inode, timespec, and timespec64 object extensions to support different kernel versions, ensuring we will get aware datetimes when using them.

gcmoreira committed:
PR review fixes: Rename method name from private to internal

gcmoreira committed:
PR review fixes: Avoid using renderers in core functions.

gcmoreira committed:
PR review fixes: Convert inode's is_* functions to properties

gcmoreira committed:
PR review fixes: Remove get_inode_number. It's better to use the type's original member name and handle the casting on the consumer side.

gcmoreira committed:
PR review fixes: Add fixme to remember we should move wintime_to_datetime/unixtime_to_datetime out of renderers

gcmoreira committed:
Linux mountinfo: Add a method to yield all filesystem superblocks

gcmoreira committed:
Linux: Improve mount's object extension method docstrings

gcmoreira committed:
Linux: dentry object extension: Add a method to walk dentries subdirectories

gcmoreira committed:
Linux: Add page cache support, including abstractions like RadixTree, XArray, and IDR, to support both older and latest kernel versions

gcmoreira committed:
Linux: Add two page cache plugins, linux.pagecache.Files and linux.pagecache.InodePages

gcmoreira committed:
Linux: Add pidhashtable plugin. This is based on the vol2 plugin, removing ancient kernel support, curating code and enhancing comments, while using the new IDR abstraction included also in this effort.

gcmoreira committed:
Linux: Add a basic eBPF program enumeration plugin to test and demonstrate using the IDR abstraction

gcmoreira committed:
Fix inode type

gcmoreira committed:
Fix @functools.cache . It's available since Python 3.9

gcmoreira commented:
Hey @ikelos, not sure why this cannot merge those changes. It looks pretty simple to me. Anyway, I'm not allowed to resolve those conflicts, it seems I will need your help here. Thanks

ikelos commented on volatility3/framework/plugins/linux/ebpf.py:
```diff
@@ -0,0 +1,78 @@
+        bpf_prog_types = vmlinux.get_enumeration("bpf_prog_type")
+        for prog in self.get_ebpf_programs(vmlinux):
+            prog_addr = prog.vol.offset
+            prog_type = bpf_prog_types.lookup(prog.type)
```
I feel like you *should* just be able to ask `prog.type.lookup()` if prog of the enumeration type?  You could also do `prog.type.description` although it's technically more characters...

gcmoreira replied:
Done. I realized I had already added a `bpf_prog` extension object that returned the prog type when I was working on `Sockstat`. So, I updated the eBPF extension objects. This change will improve support for older kernels and help avoid code duplication in future eBPF plugins.

ikelos commented on volatility3/framework/plugins/linux/mountinfo.py:
```diff
@@ -247,6 +247,31 @@ def _generator(
+        # No filter so that we get all the mount namespaces from all tasks
+        pid_filter = pslist.PsList.create_pid_filter()
+        tasks = pslist.PsList.list_tasks(
+            self.context, self.config["kernel"], filter_func=pid_filter
```
`filter_func` can be unset, and it should default to the same thing?

ikelos commented on volatility3/framework/plugins/linux/mountinfo.py:
```diff
@@ -247,6 +247,31 @@ def _generator(
+        )
+
+        seen_sb_ptr = set()
+        for task, mnt, _mnt_ns_id in self._get_tasks_mountpoints(tasks):
```
Thanks, good to know where to get the namespace if we later want to use it!  5:D

ikelos commented on volatility3/framework/plugins/linux/pagecache.py:
```diff
@@ -0,0 +1,504 @@
+vollog = logging.getLogger(__name__)
+
+
+@dataclass
```
Do these belong here, or should they live in LinuxUtilities somewhere (ie, are they always likely to be used with this plugin, or should they be more general), given that there are already elf and bash extesions under `framework/symbols/linux/extensions` so support for adding new files and common types already exists (albeit, it does require tinkering with `__init__.py` to import them be default, I think).

gcmoreira replied:
Hm I don't think so. These dataclasses are specific for this plugin and its output. The idea of these two is similar to the MVC design pattern. Being InodeInternal and InodeUser the Model and View representations respectively. 
They work great! The code becomes a lot simpler and clearer.

ikelos commented on volatility3/framework/plugins/linux/pagecache.py:
```diff
@@ -0,0 +1,504 @@
+        return symlink_path
+
+    @classmethod
+    def _walk_dentry(cls, seen_dentries, root_dentry, parent):
```
Could do with a docstring and typing please, given you've been so good about the others...

ikelos commented on volatility3/framework/plugins/linux/pagecache.py:
```diff
@@ -0,0 +1,504 @@
+        ]
+
+    @staticmethod
+    def _follow_symlink(inode, symlink_path) -> str:
```
Some typing isn't here, could we at least try to include simple types (which I'm assuming symlink_path is) please?

ikelos commented on volatility3/framework/plugins/linux/pagecache.py:
```diff
@@ -0,0 +1,504 @@
+        cls,
+        context: interfaces.context.ContextInterface,
+        config_path: str,
+    ):
```
If this yields a generator of inodes, could we capture that in the return type please?  Even if it's just `Generator[interfaces.object.ObjectInterface, None, None]`

gcmoreira replied:
Done

ikelos commented on volatility3/framework/plugins/linux/pagecache.py:
```diff
@@ -0,0 +1,504 @@
+            for header, field in zip(headers, fields):
+                header_type = header[1]
+
+                if isinstance(
```
There was a recent push to allow you to cast the response values as `format_hints.HexOrAbsent` to take care of the absent values, but that still looks like it might be a little cumbersome in this instance.  I'll look again and see what I can do about easing the type tests for these...

gcmoreira replied:
Hmm... that's interesting! I'm not sure how useful it will be here, but feel free to give it a try!

ikelos commented on volatility3/framework/plugins/linux/pagecache.py:
```diff
@@ -0,0 +1,504 @@
+            ),
+            requirements.StringRequirement(
+                name="find",
+                description="Filename (full path) to find",
```
Is there a reason this isn't a substring match, but a full path match instead?  Substring presumably encompasses both (although you wouldn't get just the first hit).

ikelos replied:
Looks like comments in the pagecache file were hidden because the changes were too big, but do please review them.

gcmoreira replied:
yeah, typically we only want to match one file. Otherwise, it has to find a way to find a name for each of them avoiding name collisions/overwriting the same file, escaping slashes, etc. I mean, not a big deal but let's keep it simple for now.

ikelos commented on volatility3/framework/plugins/linux/pidhashtable.py:
```diff
@@ -0,0 +1,249 @@
+
+    def _get_pidtype_pid(self):
+        # The pid_type enumeration is present since 2.5.37, just in case
+        pid_type_enum = self.vmlinux.get_enumeration("pid_type")
```
I'd much prefer this parameterized than maintaining state.  I can't fortell if someone's going to reuse plugins in a different UI somehow, so maintaining it by passing it through parameters rather than state seems better and would allow these functions to be exposed later on if they prove useful for others.

ikelos commented on volatility3/framework/plugins/linux/pidhashtable.py:
```diff
@@ -0,0 +1,249 @@
+
+        array_type_name = self.vmlinux.symbol_table_name + constants.BANG + "array"
+
+        pidhash_ptr = self.vmlinux.object_from_symbol("pid_hash")
```
Hopefully these methods don't get called too often, otherwise passing through simple parameters and having to do look up would be quite inefficient, but I think I'd still prefer a complex parameter to a self one in a case like this

ikelos commented on volatility3/framework/renderers/conversion.py:
```diff
@@ -11,6 +11,7 @@
 from volatility3.framework import interfaces, renderers
 
 
+# FIXME: Move wintime_to_datetime() and unixtime_to_datetime() out of renderers, possibly framework.objects.utility
```
Thanks!  5:)

ikelos commented on volatility3/framework/symbols/linux/__init__.py:
```diff
@@ -419,3 +430,342 @@ def get_module_from_volobj_type(
+            return RadixTree(vmlinux)
+
+
+class Tree(ABC):
```
The name's a little generic.  The module helps identify it, but if there's a slightly more descriptive name than Tree, it would be good to just differentiate it a little.  There are other operating systems in here too, and we don't want all of them to have their own tree (plus, I'm quite sure linux will have different tree structures too).  5;P

gcmoreira replied:
Yeah, sorry I know. Actually, 'Tree' isn't quite right since XArray is not a tree. I had a hard time coming up with a good name. Basically, it's an abstract base class which allows selecting and executing the Linux kernel RadixTree and XArray data structures. Earlier kernel versions used RadixTree in various subsystems, but many of these have been replaced with XArrays.  Since these serve various functions in the kernel, it seems best to keep the name generic either way. What do you think of something like DSAbstract or DSBase (DS => Data Structure)?, any better?

gcmoreira replied:
Ok, renamed to IDStorage as it represents what all of them store, which is identifiers.

ikelos commented on volatility3/framework/symbols/linux/__init__.py:
```diff
@@ -419,3 +430,342 @@ def get_module_from_volobj_type(
+    CHUNK_SIZE = None
+    CHUNK_MASK = None
+
+    def __init__(self, vmlinux: interfaces.context.ModuleInterface):
```
For something like this, I'd prefer to pass in the context and a name lookup for `context.modules` please.  It ensures that the state is front and center, and that the rest of the parameters are simple in case they need to be pickled or anything for multithreading down the line.

gcmoreira replied:
done

ikelos commented on volatility3/framework/symbols/linux/__init__.py:
```diff
@@ -419,3 +430,342 @@ def get_module_from_volobj_type(
+        Returns:
+            A string with the node type name
+        """
+        raise NotImplementedError()
```
Ok with these assuming they'll be filled in at some point?

gcmoreira replied:
Yes, see the RadixTree and XArray classes

ikelos commented on volatility3/framework/symbols/linux/__init__.py:
```diff
@@ -419,3 +430,342 @@ def get_module_from_volobj_type(
+            page_cache: Page cache address space
+            vmlinux: Kernel module object
+        """
+        self.vmlinux = vmlinux
```
Same as above, ideally.  First parameter would be `context`, then I'd go `vmlinux_name: str` and then `page_cache: interfaces.objects.ObjectInterface` or even `page_cache: 'PageCache'`

gcmoreira replied:
done

ikelos commented on volatility3/framework/symbols/linux/__init__.py:
```diff
@@ -419,3 +430,342 @@ def get_module_from_volobj_type(
+        """
+        self.vmlinux = vmlinux
+        self._page_cache = page_cache
+        self._tree = LinuxUtilities.choose_kernel_tree(self.vmlinux)
```
In fact, if this is all it's used for, you can just pass the `context` and `vmlinux_name` right through.

gcmoreira replied:
done

ikelos reviewed (commented):
Well, quite a long chunk of code and if this weren't from someone that already knows most of the code style points I'd raise, I'd have asked them to come back and submit them as several standalone chunks, but you've been submitting long enough you don't fall for most of the common traps so it didn't take too long to review.  5;P

Some docstrings and types in places (some are so short it's easier just to read the method, but any over a few lines should get their own docstring please, it'll help when people aren't looking at the source but just the documentation).

Towards the end I started noticing several functions that just passed an object through and in general we try hard to pass a context and then simple parameter types (like strings) that are used as a look key instead.  I've mentioned the reasons in the comments (multithreading and ease of understanding given types are hard to type further than `ObjectInterface`), but at this point it's practically code style for the project.

Generally though, it's all very nice,and I'm  very happy with some of the touches in here.  Really good work, thanks!

ikelos committed:
Merge branch 'develop' into linux_page_cache

ikelos commented:
> Hey @ikelos, not sure why this cannot merge those changes. It looks pretty simple to me. Anyway, I'm not allowed to resolve those conflicts, it seems I will need your help here. Thanks

Yeah, that was a bit weird, they weren't really conflict from what I could see, but I guess some of the recent merges touched the files in those places.  Anyway, should be all resolved, lemme know if there's an issue...

gcmoreira committed:
PR review fixes: Improve eBPF extension objects: bpf_prog and added bpf_prog_aux.  Apply changes to the EBPF and Sockstat plugins.

gcmoreira committed:
PR review fixes: Fix pidhashtable plugin

gcmoreira committed:
PR review fixes: Fix pidhashtable plugin explicit returns mixed with implicit returns

gcmoreira committed:
PR review fixes: pidhashtable plugin add missing typing.

gcmoreira committed:
PR review fixes: Remove filter function, it isn't needed

gcmoreira committed:
PR review fixes: Improve _walk_dentry() and get_inodes()  variable names, arguments and return values

gcmoreira committed:
PR review fixes: Fix page flags list method name, this was introduced earlier in another commit of this PR.

gcmoreira committed:
PR review fixes: Use contextlib.suppress() instead of an empty exception handler

gcmoreira committed:
PR review fixes: ebpf plugin code improvement. Use the object_from_symbol() instead

gcmoreira committed:
PR review fixes: Fix IDR explicit returns mixed with implicit returns and improve and fix code.

gcmoreira committed:
PR review fixes: Since the IDR, XArray and RadixTree can store any value, it renames the function names to a more generic name

gcmoreira committed:
PR review fixes: Code scanning complains about these unused variables. Let's comment them and adapt the FIXME message

gcmoreira committed:
PR review fixes: Fix the IDR's old kernel get_entries

gcmoreira committed:
PR review fixes: Fix page extension object get_content() explicit returns mixed with implicit returns.

gcmoreira committed:
linuxutilities code improvement. Remove code duplication

gcmoreira committed:
PR review fixes: Rename Tree to IDStorage. Move choose_id_storage() form LinuxUtilities to IDStorage. Use context and kernel_module_name instead of vmlinux

gcmoreira committed:
PR review fixes: Fix minor typo to match verb form from other docstrings

gcmoreira commented:
@ikelos I changed the state to `draft`. It's almost ready but I want to add some testcases for completeness. 

In the meantime, if you can please have a look at the **LinuxUtilities** version. I bumped the minor number when I added the ex **choose_kernel_tree()** helper, but, I later moved it to the **IDStorage** class. However, I made a code improvement,  which could also be considered a **patch**, since the functionality itself hasn't changed. Not sure the exact criteria we should take into account for this. If we need to update the version again, the **pidhashtable** plugin requirements also needs to be adjusted.

gcmoreira committed:
Merge branch 'volatilityfoundation:develop' into linux_page_cache

ikelos commented:
No problem moving this to draft.  If the change you've made doesn't add to the LinuxUtilities interface (or remove/change it in any way) but just alters *how* it does the stuff it does (without changing the results) then it can just be bumped on the PATCH version rather than the MINOR version.  Depends on the patch, but should be ok (otherwise we tend to leave the old functionality around with a deprecation warning, because bumping a MAJOR version on LinuxUtilities will be... painful.  5:P

ikelos commented on volatility3/framework/symbols/linux/__init__.py:
```diff
@@ -67,7 +74,7 @@ def __init__(self, *args, **kwargs) -> None:
     """Class with multiple useful linux functions."""
 
-    _version = (2, 1, 0)
+    _version = (2, 2, 0)
```
This is right, given that `IDStorage` is being added. :+1:

ikelos commented on volatility3/framework/symbols/linux/extensions/__init__.py:
```diff
@@ -1584,20 +1608,59 @@ def get_state(self):
 class bpf_prog(objects.StructType):
-    def get_type(self):
+    def get_type(self) -> Union[str, None]:
+        """Returns a string with the eBPF program type"""
```
Awesome, thanks!  The typing and docstrings will help a lot  5:)

ikelos reviewed (commented):
There's getting to be a *lot* of changes in this one PR.  I realize they're all related/linked, but it might be better to add in distinct classes as their own PRs, and then tie them all together with a plugin that uses them later on, just so the reviewer doesn't have pages and pages of code to have to check over.  It might be easier to spot smaller gotchas (although it may also be harder to see how things interact).  This is also why having the typing information helps a lot, because it means that checking cross links to things should be easier knowing what they take in/return...

gcmoreira committed:
Fix test.yaml, it should remove *.bin and not *.lime. There is no *.lime atm.

gcmoreira committed:
PR review fixes: Adjust LinuxUtilities version since we moved choose_id_storage() back to the IDStorage class.

gcmoreira commented:
Hey @ikelos, IMO it's ready to merge. Again, my apologies. I know it's quite extensive chuck of code. I tried to break it down in some PRs, but I know I could have done it even better.  I'll work on improving this for the next PR. 
On a positive note, I'm pretty sure our DFIR users will find these features very useful. Cheers

ikelos commented on volatility3/framework/plugins/linux/mountinfo.py:
```diff
@@ -247,6 +247,28 @@ def _generator(
+
+        seen_sb_ptr = set()
+        for task, mnt, _mnt_ns_id in self._get_tasks_mountpoints(tasks):
+            path_root = linux.LinuxUtilities.get_path_mnt(task, mnt)
```
Does this now need LinuxUtilities `(2,2,0)` or will it work with `(2,1,0)` as the plugin says?

gcmoreira replied:
linux.LinuxUtilities.get_path_mnt() wasn't modified by this PR. This can keep the (2,1,0) requirement

ikelos commented on volatility3/framework/plugins/linux/mountinfo.py:
```diff
@@ -247,6 +247,28 @@ def _generator(
                 "Could not filter by mount namespace id. This field is not available in this kernel."
             )
 
+    def get_superblocks(self):
```
Since this *isn't* being exposed as part of the plugin's API, it might be better to make that explicit by starting the function with `_get_superblocks` so that people don't try calling it from outside the plugin.  Either that, or you could expose it (although I'd probably parameterized it on `tasks` if you're going to do that.

gcmoreira replied:
It's exposed and used by the **pagecache.Files** plugin, it's called from **get_inodes()**. This gathers all the system superblocks. To obtain that list, it needs to walk all the tasks. However, it doesn't make sense to call this with a subset of tasks.

ikelos replied:
Ok, if it's supposed to be called from other plugins, it needs to be a classmethod or staticmethod then.  Instantiating this plugin within another plugin just for this functionality is a bit overkill and doesn't match the rest of the framework for how plugins expose functionality.

ikelos commented on volatility3/framework/plugins/linux/ebpf.py:
```diff
@@ -0,0 +1,68 @@
+            ),
+        ]
+
+    def get_ebpf_programs(self, vmlinux) -> interfaces.objects.ObjectInterface:
```
This isn't being expose, so it's *technically* not a problem, but we tend to prefer call signatures to be `context` and then `module_name` and the lookup happen inside the function.  Your call whether to update it, but if do you ever intend to expose `get_ebpf_programs` (which seem like they'd be handy) then I'd prefer you did make this change please.

gcmoreira replied:
done

ikelos commented on volatility3/framework/plugins/linux/pagecache.py:
```diff
@@ -0,0 +1,519 @@
+        # Additionally, using the page index will guarantee that each page is written at the
+        # appropriate file position.
+        try:
+            with open(filename, "wb") as f:
```
This breaks if this is actually being used through a web interface or similar.  All file operations need to be conducted using the [plugin's `open` method](https://github.com/volatilityfoundation/volatility3/blob/4578434bbd1587a8ac6c56fcc2757d3dcbe2edb7/volatility3/framework/interfaces/plugins.py#L146).  See any of the `dump` methods for an example, but essentially make a file handle using `open` and then set the preferred filename attribute on the file handle.

gcmoreira replied:
done

ikelos commented on volatility3/framework/plugins/linux/pidhashtable.py:
```diff
@@ -0,0 +1,249 @@
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.vmlinux = None
```
Saving state is extremely uncommon in plugins and may prevent us doing this with plugins in the future (certainly it prevents us exposing methods to be used by other plugins).  Ideally every method that makes use of these would be parameterized to accept them (or more accurately, the context and the name to lookup the module).

gcmoreira replied:
resolved

ikelos commented on volatility3/framework/plugins/linux/pidhashtable.py:
```diff
@@ -0,0 +1,249 @@
+        Yields:
+            task_struct objects
+        """
+        self.vmlinux = self.context.modules[self.config["kernel"]]
```
Again, this might be better parameterizing the `context` and `kernel` value, and then make it a classmethod?

gcmoreira replied:
resolved

ikelos commented on volatility3/framework/symbols/linux/extensions/__init__.py:
```diff
@@ -1897,3 +1960,250 @@ def get_file_mode(self) -> str:
+            A dictionary with the pageflags enumeration key/values
+        """
+        # FIXME: It would be even better to use @functools.cached_property instead,
+        # however, this requires Python +3.8
```
We've got a PR ( #1208 ) that may push the minimum version to 3.8, so keep an eye out.

ikelos commented on volatility3/framework/symbols/linux/extensions/__init__.py:
```diff
@@ -1897,3 +1960,250 @@ def get_file_mode(self) -> str:
+        Returns:
+            int: page physical address
+        """
+        vmlinux = linux.LinuxUtilities.get_module_from_volobj_type(self._context, self)
```
This doesn't check the version of `LinuxUtilities`, so we need to be a little careful using it.  Since these are both core components it should be ok, but it might be better to just check the version and throw an exception is it's not right (you can check the `VersionRequirement` code to see how to do that).

gcmoreira replied:
hm done, but I'm not sure we want this, have a look at the last commit

ikelos replied:
Yeah, I see what you mean, but... I would like the check in there?  It looks like it could be reduced to a static method somewhere else that then appears as a oneliner at call sites, if the length/duplicative nature is what feels ugly?  I imagine a check like that will be useful for other code (or maybe even some that's already snuck in).

gcmoreira replied:
Yeah, but it's becoming a checking versions infinite loop hehe. The get_module_from_volobj_type() was in LinuxUtilities to allow being reused, which need a version. Now, if we make another method to allow reusing the code to check the version of LinuxUtilities when calling get_module_from_volobj_type(), it won't also need a version :) ?

gcmoreira replied:
@ikelos If I didn't miss anything, this should be the only open discussion remaining. Let me know if you're happy with how I fixed it

ikelos replied:
I might've missed where this code was defined.  If it's defined *in* LinuxUtilities then it can be certain of its own version number, but otherwise it should be checking to see that it's available.  I didn't mean add another method to LinuxUtilities, I meant add it to a shareable base class, but it's fine if not.  I can sometimes try to reduce co-reuse too much...  5;P

ikelos replied:
See my comment below, since these are changes to the linux extensions, they're rolled into the main framework version number, whose MINOR number should be bumped whenever an addition is made...

ikelos replied:
(That also means we can rip out those LinuxUtilities checks I asked for and just rely on the framework and LinuxUtilities versions agreeing (sorry for the faffing, but getting these right so that people's plugins don't mysteriously not work on a unknown configurations is getting to be complex).

ikelos commented on volatility3/framework/symbols/linux/extensions/__init__.py:
```diff
@@ -1897,3 +1960,250 @@ def get_file_mode(self) -> str:
+                # !CONFIG_DYNAMIC_MEMORY_LAYOUT
+                if vmlinux_layer._maxvirtaddr < 57:
+                    # 4-Level paging -> VMEMMAP_START = __VMEMMAP_BASE_L4
+                    vmemmap_base_l4 = 0xFFFFEA0000000000
```
Mmmmm, not keen on hard coded values in here.  I assume this can't be calculated from any other values you can get?

gcmoreira replied:
Yeah, I agree, but yeah it's a kernel constant, so I kept the kernel constant names for clarity. It also shouldn't be used anywhere outside of this context. Otherwise, I would have put it in the Linux constants file.

ikelos reviewed (commented):
Ok, well.  There's a *lot* to go through there, so get that fixed up, I'll give it another quick pass and then we'll need to keep an eye on it and see if anything breaks...  5:S  Ideally smaller chunks in future please...

gcmoreira committed:
PR review fixes: Use plugin's open method instead of the builtin open()

gcmoreira committed:
PR review fixes: Use context and module_name instead of vmlinux in ebpf plugin

gcmoreira committed:
PR review fixes: Avoid saving state in the pidhashtable plugin

gcmoreira committed:
PR review fixes: Check for LinuxUtilities version everywhere we use it

gcmoreira commented:
@ikelos if you are happy with the last changes, it should be ready to go

ikelos commented on volatility3/framework/plugins/linux/pagecache.py:
```diff
@@ -0,0 +1,521 @@
+            An InodeInternal object
+        """
+
+        superblocks_iter = mountinfo.MountInfo(
```
This is a lot of machinery to call a function in another plugin.  Ideally the other plugin exposes it as a classmethod, and then there's no construction of the plugin required at all.

gcmoreira replied:
done

ikelos commented on volatility3/framework/symbols/linux/extensions/__init__.py:
```diff
@@ -820,6 +823,26 @@ def d_ancestor(self, ancestor_dentry):
             current_dentry = current_dentry.d_parent
         return None
 
+    def get_subdirs(self) -> interfaces.objects.ObjectInterface:
```
I feel as though the addition of this method should require a bump to the PATCH number at least, if not the MINOR number.  The only question is whether it's the whole framework or if this is part of the LinuxUtilities which is already versioned separately...

gcmoreira replied:
This is in the `dentry` extension object, but I don't think we've ever changed any versions when there were changes in this area. Hmm, it is not related to `LinuxUtilities`.

ikelos replied:
Ok, I think I see what's going on here (and the LinuxUtilities checks I asked for earlier).  We've got bits of the core framework (extensions) that aren't versioned other than under the umbrella of the main core, that are using LinuxUtilities which we gave its own version number to.  So what *should* be/have been happening is the main core bumps (at least PATCH but probably minor) whenever a linux extension was added, only I'm not sure we always did that because people hate screwing with version numbers...  5:S  Probably we can fix this by going up by making sure all the extensions we want to add are in place, and then going to (2,9,0) and all plugins that need/make use of it require that as a minimum framework version.  Hopefully that'll clear everything up and even if it's a little broken in the middle, it'll give people a new version number they can then rely on to have the functionality they want/need.

ikelos reviewed (commented):
Github keeps making this look smaller than it is by hiding files that changed a lot by default!  It's an enourmous patch and in future I'd prefer the groundwork/linux core stuff goes in first, then the plugins going in one by one (or at least file by file).  Having said that, this now looks just about done, but do please check *all* the comments (even those that github hides) to make sure they don't get missed.

gcmoreira committed:
PR review fixes: Add typing info to the get_inodes() class method.

gcmoreira committed:
PR review fixes: Make mountinfo.get_superblocks() a classmethod and adapt the code using it.

gcmoreira committed:
PR review fixes: Add typing info to pagecache.Files._follow_symlink()

gcmoreira commented:
@ikelos I believe everything is done on my end. Please take a look when you have time. Cheers

ikelos commented on volatility3/framework/plugins/linux/pidhashtable.py:
```diff
@@ -44,17 +44,16 @@ def get_requirements(cls) -> List[interfaces.configuration.RequirementInterface]
-        return bool(task and task.pid > 0 and self.vmlinux_layer.is_valid(task.parent))
+        vmlinux = self.context.modules[self.config["kernel"]]
+        vmlinux_layer = self.context.layers[vmlinux.layer_name]
+        return bool(task and task.pid > 0 and vmlinux_layer.is_valid(task.parent))
```
Why not just use the task's layer rather than the vmlinux layer (or has the task been made in a physical or process layer)?  `task.vol.layer_name` should get you back to the layer the task was created on?

ikelos commented on volatility3/framework/plugins/linux/pidhashtable.py:
```diff
@@ -44,17 +44,16 @@ def get_requirements(cls) -> List[interfaces.configuration.RequirementInterface]
+        vmlinux_layer = self.context.layers[vmlinux.layer_name]
+        return bool(task and task.pid > 0 and vmlinux_layer.is_valid(task.parent))
 
     def _get_pidtype_pid(self):
```
I have to admit, I thought these would end up being parameterized classmethods, but as long as you're happy that they're only for use in this plugin this is fine.

ikelos commented:
Yep, looks good, going to merge now, thanks for your patience getting it all sorted!  5:)

gcmoreira commented:
> Yep, looks good, going to merge now, thanks for your patience getting it all sorted! 5:)

Thank you, and again... very sorry for the long PR, I will submit smaller PRs next time. 
Very happy that Volatility3 now includes these features
